common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -409,4 +410,25 @@ private void appendUnit(StringBuilder sb, long value, String unit) {
       sb.append(' ').append(value).append(' ').append(unit).append('s');
     }
   }
+
+  /**
+   * Gets interval duration
+   * @param daysPerMonth the number of days per one month
+   * @param targetUnit time units of the result
+   * @return duration in the specified time units
+   */
+  public long getDuration(int daysPerMonth, TimeUnit targetUnit) {
+    long monthsDuration = Math.multiplyExact(daysPerMonth * MICROS_PER_DAY, months);
+    long result = Math.addExact(microseconds, monthsDuration);
+    return targetUnit.convert(result, TimeUnit.MICROSECONDS);
+  }
+
+  /**
+   * Checks the interval is negative
+   * @param daysPerMonth the number of days per one month
+   * @return true if duration of the given interval is less than 0 otherwise false
+   */
+  public boolean isNegative(int daysPerMonth) {
```

Comment:
Is it necessary to expose this vs just letting callers get the duration and compare?

Reply:
I added it for code readability at caller side.

Reply:
Also it could be considered as a companion method for [negate()](https://github.com/apache/spark/blob/da576a737c2db01e5ba5ce19ed0e8f900cb5efaf/common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java#L358-L360)

Reply:
@srowen Do you think it is not worth it?

Reply:
I'm neutral on it. It makes some sense; I'm just always reluctant to add API methods.

Reply:
Hm, on a second look:
- Can this be static? 
- Can the methods be package private?
- Why does the caller define how many days are in a month? would that assumption not be fixed?
- More generally, do we really need to allow for negative intervals at all? semantically it's nonnegative. Is it allowed in SQL?

Reply:
> Can this be static?

Yes, it can.

> Can the methods be package private?

Probably, yes.

> Why does the caller define how many days are in a month?

Because number of days per month is not constant. Structure Streaming follows conservative approach, and assumes 31 days per months (see https://github.com/apache/spark/pull/16304#discussion_r92753590). In another places, we can assume another numbers (see https://github.com/apache/spark/blob/77fe8a8e7cdb0eae9f66b2d10c0d35f784663547/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala#L31).

> would that assumption not be fixed?

For now, the methods are used by SS only. Apparently 31 can be hard coded inside.

> More generally, do we really need to allow for negative intervals at all? semantically it's nonnegative. Is it allowed in SQL?

The SQL standard allows negative intervals. If we look at other DMBS, PostgreSQL allows as well. If you subtract 2 timestamp columns `timestamp1 - timestamp2`, what are you going to produce when `timestamp1 < timstamp2`?

Reply:
Got it, yes makes sense about negative intervals as the necessary result of subtracting timestamps.

Reply:
The `CalendarInterval` class has many static methods that can be move out of it to a separate class, for instance `CalendarIntervalUtils`. I would leave only `milliseconds()`, `add()`, `subtract()`, `negate()`, `hashCode()`, `toString()` in `CalendarInterval`, and put `fromString()`, `fromCaseInsensitiveString()`, `fromYearMonthString()` and etc. to `CalendarIntervalUtils`. 

... or any kind of refactoring is impossible in Spark already because of commit history tracking problem? @cloud-fan WDYT?

Reply:
I wouldn't bother with a big refactoring here. Whatever is consistent. But yeah you can make it static at least?

Reply:
I will move those 2 methods to `IntervalUtils`